feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts

```diff
@@ -1,155 +1,163 @@
+   * Creates unique key for attribute 'aria-labelledby'.
+   *
+   * @param {string} prefix - Prefix
+   * @param {string} attributeId - Attribute ID
+   * @param {string} valueId - Value ID
+   * @param {boolean} hasQuantity - Has quantity
+   * @return {string} - Generated UI key for 'aria-labelledby'
    */
   createAriaLabelledBy(
     prefix: string,
     attributeId: string,
     valueId?: string,
     hasQuantity?: boolean
   ): string {
-    let attributeUiKey = this.createAttributeUiKey(
-      ConfiguratorAttributeBaseComponent.PREFIX_LABEL,
-      attributeId
+    return this.uiKeyGeneratorService.createAriaLabelledBy(
+      prefix,
+      attributeId,
+      valueId,
+      hasQuantity
     );
-    if (valueId) {
-      attributeUiKey +=
-        ' ' +
-        this.createAttributeUiKey(prefix, attributeId) +
-        ConfiguratorAttributeBaseComponent.SEPERATOR +
-        valueId +
-        ' ';
-      if (typeof hasQuantity === 'boolean' && !hasQuantity) {
-        attributeUiKey +=
-          this.createAttributeUiKey(
-            ConfiguratorAttributeBaseComponent.PREFIX_DDLB_OPTION_PRICE_VALUE,
-            attributeId
-          ) +
-          ConfiguratorAttributeBaseComponent.SEPERATOR +
-          valueId;
-      } else {
-        attributeUiKey +=
-          this.createAttributeUiKey(
-            ConfiguratorAttributeBaseComponent.PREFIX_OPTION_PRICE_VALUE,
-            attributeId
-          ) +
-          ConfiguratorAttributeBaseComponent.SEPERATOR +
-          valueId;
-      }
-    }
-    return attributeUiKey;
   }
 
   /**
-   * Creates a unique key for focus handling for the given attribute and value
-   * @param attributeId
-   * @param valueCode
-   * @returns focus key
+   * Creates a unique key for focus handling for the given attribute and value.
+   *
+   * @param {string} attributeId - Attribute ID
+   * @param {string} valueCode - Value Code
+   * @return {string} - Generated focus key
    */
   createFocusId(attributeId: string, valueCode: string): string {
-    return `${attributeId}--${valueCode}--focus`;
+    return this.uiKeyGeneratorService.createFocusId(attributeId, valueCode);
   }
 
   /**
    * Get code from attribute.
    * The code is not a mandatory attribute (since not available for VC flavour),
    * still it is mandatory in the context of CPQ. Calling this method therefore only
    * makes sense when CPQ is active. In case the method is called in the wrong context, an exception will
-   * be thrown
+   * be thrown.
    *
-   * @param {Configurator.Attribute} Attribute
+   * @param {Configurator.Attribute} attribute - Attribute
    * @returns {number} Attribute code
    */
-  protected getAttributeCode(attribute: Configurator.Attribute): number {
-    const code = attribute.attrCode;
-    if (code) {
-      return code;
-    } else {
-      throw new Error('No attribute code for: ' + attribute.name);
-    }
+  getAttributeCode(attribute: Configurator.Attribute): number {
```

Comment:
we can remove the parameter `attribute` and use `this.attribute`. Please also check other functions to remove the parameters which are already defined in the component.

Reply:
Done!